spec/java/util/OptionalInt.lsl

```diff
@@ -0,0 +1,282 @@
+)
+@public automaton OptionalInt: int
+(
+    var value: int = 0;
+    var present: boolean = false;
+)
+{
+    // states and shifts
+
+    initstate Allocated;
+    state Initialized;
+
+    shift Allocated -> Initialized by [
+        // constructors
+        OptionalInt (),
+        OptionalInt (int),
+
+        // static methods
+        empty,
+        of,
+        ofNullable,
+    ];
+
+    shift Initialized -> self by [
+        // read operations
+        getAsInt,
+        isPresent,
+        isEmpty,
+        ifPresent,
+        ifPresentOrElse,
+        filter,
+        map,
+        flatMap,
+        or,
+        stream,
+        orElse,
+        orElseGet,
+        orElseThrow (),
+        orElseThrow (Supplier),
+        toString,
+        hashCode,
+        equals,
+    ];
+
+
+    // constructors
+
+    @private constructor OptionalInt ()
+    {
+        action ERROR("Private constructor call");
+    }
+
+
+    @private constructor OptionalInt (x: int)
+    {
+        action ERROR("Private constructor call");
+    }
+
+
+    // utilities
+
+    @CacheStaticOnce
+    @static proc _makeEmpty (): OptionalInt
+    {
+        // #problem
+        result = new OptionalInt(state=Initialized);
+    }
+
+
+    @AutoInline
+    proc _throwNPE (): void
+    {
+        action THROW_NEW("java.lang.NullPointerException", []);
+    }
+
+
+    // static methods
+
+    @static fun empty (): OptionalInt
+    {
+        result = _makeEmpty();
+    }
+
+
+    @static fun of (x: int): OptionalInt
+    {
+        result = new OptionalInt(state=Initialized, value=x, present=true);
+    }
+
+
+    // methods
+
+    fun equals (other: Object): boolean
+    {
+        if (other == self)
+        {
+            result = true;
+        }
+        else
+        {
+            val isSameType = action OBJECT_SAME_TYPE(self, other);
+            if (isSameType)
+            {
+                val otherValue = OptionalInt(other).value;
+                val otherPresent = OptionalInt(other).present;
+                result = self.value == otherValue && self.present == otherPresent;
+            }
+            else
+            {
+                result = false;
+            }
+        }
+    }
+
+
+    fun getAsInt (): int
+    {
+        if (!present)
+            action THROW_NEW("java.util.NoSuchElementException", ["No value present"]);
+
+        result = value;
+    }
+
+
+    fun hashCode (): int
+    {
+        result = action OBJECT_HASH_CODE(value);
+    }
+
+
+    fun ifPresent (consumer: IntConsumer): void
+    {
+        required !present || (present && consumer != null);
+
+        if (present)
+        {
+            if (consumer == null)
+                self._throwNPE();
+
+            action CALL(consumer, [value]);
+        }
+    }
+
+
+    fun ifPresentOrElse (consumer: IntConsumer, emptyAction: Runnable): void
+    {
+        required !present || (present  && consumer != null);
+        required present  || (!present && emptyAction != null);
```

Comment:
Maybe we don't need "present" in the second required ? Because we can conclude from the first required that this can be such: "required emptyAction != null" ?
Or not ?

Reply:
If the value is present, then `emptyAction` can be `null`.
If we leave just `required emptyAction != null` then it would be impossible to supply `null` as `emptyAction`.
The requirements, I suppose, are all connected using `AND` operation.
So I don't see the problem here.